Mixmatch/streetview_dataset/parse_data_to_tfrecord_lib.py

```diff
@@ -0,0 +1,235 @@
+                target = strip_all_qualified_bbox(confidence, bbox, CONF_THRESHOLD)
+            
+            if not target:
+                continue
+
+            for t in target:
+                crop_img = img
+                if 'bbox' in t:
+                    crop_img = crop_img.crop(t['bbox'])
+                crop_img = crop_img.resize(OUTPUT_IMAGE_SIZE)
+                example = img_to_example(crop_img, t['label'])
+                examples[t['label']].append(example)
+                
+    return examples
+
+# Write positive and negative tfexamples to tfrecord using the writer. A balance boolean parameter can decide to balance the pos and neg examples count.
+def write_tfexample_to_tfrecord(positive_examples, negative_examples, balance, writer):
+    take = float('inf')
+    num_pos = len(positive_examples)
+    num_neg = len(negative_examples)
+    
+    if balance:    
+        take = min(num_pos, num_neg)
+        positive_examples = positive_examples[:take]
+        negative_examples = negative_examples[:take]
+    
+    examples = positive_examples.extend(negative_examples)
+    
+    random.seed(1)
+    random.shuffle(examples)
+    
+    for i, example in enumerate(examples):
+        writer.write(example.SerializeToString())
+
+
+# Write tfrecords in batches of input record files.
+def batch_read_write_tfrecords(file_range, input_record_path, input_img_path, writer, detection_property):
+    include_top_camera = detection_property['include_top_camera']
+    only_keep_top_confidence = detection_property['only_keep_top_confidence']
+    balance = detection_property['balance']
+    pos_examples, neg_examples = [], []
+    
+    for i in range(file_range[0], file_range[1]):
+        file_name = "./streetlearn_detections_tfexample-" + str(i).zfill(5) + "-of-01000.tfrecord"
+        parsed_image_dataset = read_tfrecord(os.path.join(input_record_path, file_name))
+        examples = generate_tfexamples_from_detections(parsed_image_dataset, input_img_path, include_top_camera, only_keep_top_confidence)
+        neg_examples.extend(examples[0])
+        pos_examples.extend(examples[1])
+    
+    write_tfexample_to_tfrecord(pos_examples, neg_examples, balance, writer)
+
+
+# Filter the dataset with images bbox lower than the threshold, and copy the image bbox to output directory. These images will be handpicked to be used as negative examples in the test set.
+def filter_image_with_confidence_threshold(parsed_image_dataset, input_folder_path, output_folder_path, neg_threshold):
+    for image_features in parsed_image_dataset:
+        img_name, confidence, bbox = parse_detection_confidences(image_features)
+        # The format fo the image_name is XXXXXX_Y.jpg, the Y is the identifier of the view. 1, 2, 3 and 4 are the side views and 5 is the upward view. 0 is the view with markers overlaid.
+        view = img_name.split('.')[0][-1]
+        keep_image = view != '0' and view != '5'
+        
+        if img_name and keep_image:
```

Comment:
Prefer using early return for invalid case, i.e.
```
if not img_name or not keep_image:
  continue
```

Reply:
Done both at this function and at generate_tfexamples_from_detections().